### ISSUE TITLE
prefer-const and for loop/block scope `let` semantics

### ISSUE THREAD

**Issue by jcrugzz** [eslint/eslint#4548]
I am currently using `eslint@1.10.1` and I wanted to clarify `prefer-const` in regards to `let` scope semantics. I have a snippet similar to the following which errors with `"prefer-const": 2`. The errors occur with `scope`, `endpoint`, `stream` within the `if` statement block.

```js
function (data, enc, callback) {
  const name = data.name;

  const endpoint = this.lookups.endpoint.get('default');
  const stream = this.streams.get(endpoint);
  if (stream) stream.write(data);

  if (this.lookups.packages.has(name) {
    let scope = this.lookups.packages.get(scope);
    let endpoint = this.lookups.endpoint(scope);
    let stream = this.lookups.get(endpoint);
    if (stream) stream.write(data);
  }
// possibly do other things with function scoped variables
}
```

My assumption is that `prefer-const` would not error on `scope` `endpoint` or `stream` because of the block scope semantics being important enough to override the default behavior of erroring by using `let` where something doesn't change. At the very least, endpoint and stream shouldn't error since this it does not modify outer scope variables.

Thoughts around this? I know assumptions can be tricky with the semantic behavior of block scoping so I wanted to get some opinions around this. Thanks!

**Comment by eslintbot** [comment]
@jcrugzz Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Can you provide exact output of linting? And also which version you are using.

**Comment by nzakas** [comment]
This rule doesn't care about block scope. The only thing that matters is that you're using `let`, and the binding is never modified. For that reason, the rule tells you to use `const`.
